thali/thalireplicationmanager.js

```diff
@@ -38,25 +47,54 @@ ThaliReplicationManager.events = {
 };
 
 /**
-* Starts the Thali replication manager with the given device name and port number
-* @param {String} deviceName the device name to advertise.
+* Returns the current device's id if available. If not, the provided callback
+* function is saved and called when the id does become available.
+* @param {Function} cb the callback which returns the current device's id.
+*/
+ThaliReplicationManager.prototype.getDeviceIdentity = function (cb) {
+  if(this._deviceIdentityFlag == deviceIdentityFlag.deviceIdentityAvailable) {
+    cb(null, this._deviceName);
+  } else if(this._deviceIdentityFlag == deviceIdentityFlag.noDeviceIdentitySet) {
+    var cryptomanager = require('./thalicryptomanager');
+    this._deviceIdentityFlag = deviceIdentityFlag.gettingDeviceIdentity;
+    cryptomanager.getPublicKeyHash(function (err, publicKeyHash) {
+      if (err) {
+        this._deviceIdentityFlag = deviceIdentityFlag.noDeviceIdentitySet;
+        informDeviceIdentityListeners.call(this, err, null);
+        clearDeviceIdentityListeners.call(this);
+        cb(err);
+      } else {
+        this._deviceName = publicKeyHash;
+        this._deviceIdentityFlag = deviceIdentityFlag.deviceIdentityAvailable;
+        informDeviceIdentityListeners.call(this, null, publicKeyHash);
+        clearDeviceIdentityListeners.call(this);
+        cb(null, publicKeyHash);
+      }
+    }.bind(this));
+  } else {
+    // save the callback for future
+    this._deviceIdentityListeners.push(cb);
+  }
+}
+
+/**
+* Starts the Thali replication manager with the given port number and db name.
+* The device-id is obtained using the cryptomanager's API.
 * @param {Number} port the port number used for synchronization.
+* @param {String} dbName the db name to sync.
 */
-ThaliReplicationManager.prototype.start = function (deviceName, port, dbName) {
-  validations.ensureNonNullOrEmptyString(deviceName, 'deviceName');
+ThaliReplicationManager.prototype.start = function (port, dbName) {
   validations.ensureValidPort(port);
   validations.ensureNonNullOrEmptyString(dbName, 'dbName');
-
   this.emit(ThaliReplicationManager.events.STARTING);
-
+  
   this._port = port;
-  this._deviceName = deviceName;
   this._dbName = dbName;
   this._serverBridge = muxServerBridge.call(this, port);
   this._serverBridge.listen(function () {
     this._serverBridgePort = this._serverBridge.address().port;
 
-    this._emitter.startBroadcasting(deviceName, this._serverBridgePort, function (err) {
+    this._emitter.startBroadcasting(this._deviceName, this._serverBridgePort, function (err) {
```

Comment:
Fixed by building getDeviceIdentity if user has not done it before calling start.